### PR TITLE
Additional ways to interact with extensions

### PR DESCRIPTION
**Pull request by new-maxim724**
As I wrote in the discord, this will help to decouple the implementation and only give access to the abstraction. It also saves us from casting the type to the required one.

Why am I not using Class<? extends Extension> ? Since we cannot inherit the interface from Extension, we have to do this.
On the one hand, this is not entirely secure, but on the other hand, we know that the ExtensionManager only loads Extension subclasses.

As example:
image

### CODE REVIEW

File: src/main/java/net/minestom/server/extensions/ExtensionManager.java [hunk 1]
```diff
@@ -612,10 +612,23 @@ public Extension getExtension(@NotNull String name) {
         return extensions.get(name.toLowerCase());
     }
 
+    @Nullable
+    @SuppressWarnings("unchecked")
+    public <E> E getExtension(@NotNull Class<E> type) {
+        return (E) extensions.values().stream()
+                .filter(extension -> extension.getClass().isAssignableFrom(type))
```
Comment (jglrxavpok) [inline src/main/java/net/minestom/server/extensions/ExtensionManager.java:619]: This won't work if different extensions try to access the same types: 
Extension A will have a different value when loading 'TestExtension.class' than the value Extension B will have.

File: src/main/java/net/minestom/server/extensions/ExtensionManager.java [hunk 2]
```diff
@@ -612,10 +612,23 @@ public Extension getExtension(@NotNull String name) {
...
+                .findAny().orElse(null);
+    }
+
     public boolean hasExtension(@NotNull String name) {
         return extensions.containsKey(name);
     }
 
+    public boolean hasExtension(@NotNull Class<?> type) {
+        return extensions.values().stream()
+                .anyMatch(extension -> extension.getClass().isAssignableFrom(type));
```
Comment (jglrxavpok) [inline src/main/java/net/minestom/server/extensions/ExtensionManager.java:629]: Same goes.

**Review by jglrxavpok — CHANGES_REQUESTED** [review]
Probably won't work as-is due to different extensions having separate extensions.
Have you checked in "real" setting with separate jars for two extensions and a third jar with a Minestom implementation?

Extension A will have a different value when loading 'TestExtension.class' than the value Extension B will have.

**Comment by mworzala** [comment]
I'm not really sure the goal here, but if it is still relevant after #537 please re open as an issue or PR to that branch. In the mean time, closing due to inactivity and in favor of #537

**Pull request closed without merging**